File: 프로그래머스/lv2/49994. 방문 길이/방문 길이.py
def solution(dirs):
    dx = [-1,1,0,0]
    dy = [0,0,1,-1]
    dic = {"U":0,"D":1,"R":2,"L":3}
    answer = 0
    q = []
    x,y = 5,5
    
    for i in dirs:
        d = dic[i]
        nx,ny = x+dx[d],y+dy[d]
        if 0<=nx<=10 and 0<=ny<=10:
            q.append((x,y,nx,ny))
            q.append((nx,ny,x,y))
            x,y = nx,ny
            
    answer = len(set(q))//2
    return answer

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    public int solution(String dirs) {
        Set<String> visited = new HashSet<>();
        int n = dirs.length();
        int x = 0;
        int y = 0;
        int[] dx = {-1,1,0,0}; // L, R, U, D
        int[] dy = {0,0,1,-1};
        
        for(int i=0; i<n; i++){
            int tx = x;
            int ty = y;
            if(dirs.charAt(i) == 'L'){
                tx += dx[0];
                ty += dy[0];
            } else if(dirs.charAt(i) == 'R'){
                tx += dx[1];
                ty += dy[1];
            } else if(dirs.charAt(i) == 'U'){
                tx += dx[2];
                ty += dy[2];
            } else if(dirs.charAt(i) == 'D'){
                tx += dx[3];
                ty += dy[3];
            }
            if (Math.abs(tx) > 5 || Math.abs(ty) > 5) continue;
            // 이전 좌표랑 현 좌표 둘 다 저장, 순서 바뀐 경우 예외
            if (!visited.contains(tx+","+ty+","+x+","+y)){
                visited.add(x+","+y+","+tx+","+ty); 
            }
            x = tx; 
            y = ty;
        }
        
        return visited.size();
    }
}